handler/mosn/host.go

```diff
@@ -0,0 +1,121 @@
+package mosn
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+	"strconv"
+
+	"mosn.io/mosn/pkg/types"
+	"mosn.io/mosn/pkg/variable"
+	"mosn.io/pkg/header"
+
+	"github.com/http-wasm/http-wasm-host-go/api/handler"
+	internalhandler "github.com/http-wasm/http-wasm-host-go/internal/handler"
+)
+
+var _ handler.Host = (*host)(nil)
+
+type host struct {
+}
+
+func (h host) GetProtocolVersion(ctx context.Context) string {
+	p, _ := variable.GetString(ctx, types.VarProtocol)
+	switch p {
+	case "Http1":
+		return "HTTP/1.1"
+	case "Http2":
+		return "HTTP/2.0"
+	}
+	return p
+}
+
+func (host) GetRequestBody(ctx context.Context) []byte {
+	return filterFromContext(ctx).reqBody.Bytes()
+}
+
+func (host) SetRequestBody(ctx context.Context, body []byte) {
+	buf := filterFromContext(ctx).reqBody
+	buf.Reset()
+	_ = buf.Append(body)
+}
+
+func (host) EnableFeatures(ctx context.Context, features handler.Features) handler.Features {
+	if f, ok := ctx.Value(filterKey{}).(*filter); ok {
+		f.enableFeatures(features)
+		return f.features
+	} else if i, ok := ctx.Value(internalhandler.InitStateKey{}).(*internalhandler.InitState); ok {
+		i.Features = i.Features.WithEnabled(features)
+		return i.Features
+	} else {
+		panic("unexpected context state")
+	}
+}
+
+func (host) GetURI(ctx context.Context) string {
+	// TODO(anuraaga): There is also variable.GetProtocolResource(ctx, api.URI), unclear if they must be kept in sync.
+	p, _ := variable.GetString(ctx, types.VarPath)
+	q, qErr := variable.GetString(ctx, types.VarQueryString)
+	if qErr != nil {
+		// No query, so an error is returned.
+		return p
+	}
+	return fmt.Sprintf("%s?%s", p, q)
+}
+
+func (host) SetURI(ctx context.Context, uri string) {
+	u, err := url.ParseRequestURI(uri)
+	if err != nil {
+		panic(err)
+	}
+	_ = variable.SetString(ctx, types.VarPath, u.Path)
+	if len(u.RawQuery) > 0 || u.ForceQuery {
+		_ = variable.SetString(ctx, types.VarQueryString, u.RawQuery)
+	}
+}
+
+func (host) GetRequestHeader(ctx context.Context, name string) (string, bool) {
+	return filterFromContext(ctx).reqHeaders.Get(name)
+}
+
+func (host) SetResponseHeader(ctx context.Context, name, value string) {
+	hdrs := filterFromContext(ctx).respHeaders
+	if hdrs == nil {
+		hdrs = header.CommonHeader(make(map[string]string))
+		filterFromContext(ctx).respHeaders = hdrs
+	}
+	hdrs.Set(name, value)
+}
+
+func (host) Next(ctx context.Context) {
+	f := filterFromContext(ctx)
+	f.nextCalled = true
+	f.ch <- nil
```

Comment:
was this a relic? if not, comment?

Reply:
Good call it's relatively confusing. I added a bunch of comments, do they make sense?

Reply:
yeah looks good!